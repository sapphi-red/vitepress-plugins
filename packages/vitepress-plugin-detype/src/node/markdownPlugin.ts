import type MarkdownIt from 'markdown-it'
import { transform, removeMagicComments } from 'detype'
import type { PrettierOptions } from 'detype'
import type { ContentMap } from './contentMap'
import { parseDetypeInfo, SupportedType } from './parseDetypeInfo'

const tabsShareStateKey = '~detype'
const langs = ['ts', 'js'] as const
type Lang = typeof langs[number]

const getLangForRender = (type: SupportedType, lang: Lang) => {
  switch (type) {
    case 'vue':
      return 'vue'
    case 'tsx':
      return `${lang}x`
    case 'ts':
      return lang
  }
  exhaustiveCheck(type)
}

// eslint-disable-next-line @typescript-eslint/no-empty-function
const exhaustiveCheck = (_p: never) => {}

export const detypePlugin = (
  md: MarkdownIt,
  prettierOptions: PrettierOptions,
  contentMap: ContentMap
) => {
  const tabLabelsProp = `:tabLabels="${md.utils.escapeHtml(
    JSON.stringify(langs)
  )}"`
  const shareStateKeyProp = `sharedStateKey="${md.utils.escapeHtml(
    tabsShareStateKey
  )}"`

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const originalFence = md.renderer.rules.fence!
  md.renderer.rules.fence = (...args) => {
    const [tokens, idx, , env] = args
    const token = tokens[idx]

    const parsed = parseDetypeInfo(token.info)
    if (parsed === null) {
      return originalFence(...args)
    }

    const { type, tsAttrs, jsAttrs } = parsed

    const slots = langs.map(lang => {
      const attrs = lang === 'ts' ? tsAttrs : jsAttrs
      const output = (async () => {
        try {
          const content =
            lang === 'ts'
              ? removeMagicComments(
                  token.content,
                  `foo.${type}`,
                  prettierOptions
                )
              : await transform(token.content, `foo.${type}`, {
                  prettierOptions,
                  removeTsComments: true
                })
          const langForRender = getLangForRender(type, lang)
          const codeWithFence =
            `${token.markup}${langForRender}${attrs}\n` + content + token.markup
          return { result: md.render(codeWithFence, env) }
        } catch (e) {
          return { error: e }
        }
      })()
      const key = contentMap.add(
        `${type}_${token.markup}_${lang}_${attrs}`,
        token.content,
        output
      )
      return `<template #${lang}>${key}</template>`
    })

    return `<PluginTabs ${tabLabelsProp} ${shareStateKeyProp}>${slots.join(
      ''
    )}</PluginTabs>`
  }
}
