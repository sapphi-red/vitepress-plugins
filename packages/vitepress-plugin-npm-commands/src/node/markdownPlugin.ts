import type MarkdownIt from 'markdown-it'
import { autoConverter } from './autoConverter'
import type { PackageManager } from './packageManager'
import { packageManagers } from './packageManager'

const tabsShareStateKey = '~npm-commands'
const npmCommandsCommandRE = new RegExp(
  `(?://\\s*\\[!=npm\\s+(${packageManagers.join('|')}|auto)\\])+$`,
  'm'
)

export const npmCommandsPlugin = (md: MarkdownIt) => {
  const shareStateKeyProp = `sharedStateKey="${md.utils.escapeHtml(
    tabsShareStateKey
  )}"`

  const originalFence = md.renderer.rules.fence!
  md.renderer.rules.fence = (...args) => {
    const [tokens, idx, , env] = args
    const token = tokens[idx]
    if (token.attrGet('----npm-disable') !== null) {
      // escape hatch to disable this feature
      const pos = token.attrIndex('----npm-disable')
      token.attrs?.splice(pos, 1)
      return originalFence(...args)
    }
    if (!npmCommandsCommandRE.test(token.content)) {
      return originalFence(...args)
    }

    const codes = generateEachPackageManagerCode(token.content)

    const slots = codes.map(([pkgManger, code]) => {
      const attrStr = token.attrs
        ? `{${token.attrs
            .map(([key, val]) => (val ? `${key}=${val}` : key))
            .join(' ')}}`
        : ''
      const codeWithFence =
        `${token.markup}${token.info}${attrStr}\n` + code + token.markup
      return `<PluginTabsTab label="${pkgManger}">${md.render(
        codeWithFence,
        env
      )}</PluginTabsTab>`
    })

    return `<PluginTabs ${shareStateKeyProp}>${slots.join('')}</PluginTabs>`
  }
}

const parseNpmCommandsCommandFromLine = (line: string) => {
  const m = line.match(npmCommandsCommandRE)
  if (!m) return null
  return m[1] as PackageManager | 'auto'
}

const generateEachPackageManagerCode = (input: string) => {
  const codes = Object.fromEntries(
    packageManagers.map(m => [m, [] as string[]])
  ) as Record<PackageManager, string[]>
  for (const line of input.split('\n')) {
    const pkgManager = parseNpmCommandsCommandFromLine(line)
    if (!pkgManager) {
      for (const key of packageManagers) {
        codes[key].push(line)
      }
    } else if (pkgManager === 'auto') {
      for (const key of packageManagers) {
        const convertedCommand = autoConverter(
          line.replace(npmCommandsCommandRE, '').trimEnd(),
          key
        )
        codes[key].push(convertedCommand)
      }
    } else {
      codes[pkgManager].push(line.replace(npmCommandsCommandRE, ''))
    }
  }
  return Object.entries(codes).map(
    ([key, val]) => [key as PackageManager, val.join('\n')] as const
  )
}
