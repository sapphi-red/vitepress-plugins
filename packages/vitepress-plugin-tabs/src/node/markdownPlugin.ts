import type MarkdownIt from 'markdown-it'
import container from 'markdown-it-container'
import type Token from 'markdown-it/lib/token'

type Params = {
  shareStateKey: string | undefined
}

const parseTabsParams = (input: string): Params => {
  const match = input.match(/key:(\S+)/)
  return {
    shareStateKey: match?.[1]
  }
}

export const tabsPlugin = (md: MarkdownIt) => {
  md.use(container, 'tabs', {
    render(tokens: Token[], index: number) {
      const token = tokens[index]
      if (token.nesting === 1) {
        const params = parseTabsParams(token.info)
        const shareStateKeyProp = params.shareStateKey
          ? `sharedStateKey="${md.utils.escapeHtml(params.shareStateKey)}"`
          : ''
        return `<PluginTabs ${shareStateKeyProp}>\n`
      } else {
        return `</PluginTabs>\n`
      }
    }
  })

  md.use(container, 'tab', {
    render(tokens: Token[], index: number) {
      const token = tokens[index]
      if (token.nesting === 1) {
        const label = token.info.replace(/^\s*tab\s*/, '')
        const labelProp = `label="${md.utils.escapeHtml(label)}"`
        return `<PluginTabsTab ${labelProp}>\n`
      } else {
        return `</PluginTabsTab>\n`
      }
    }
  })
}
