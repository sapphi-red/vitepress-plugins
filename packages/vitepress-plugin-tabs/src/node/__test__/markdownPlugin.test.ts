import MarkdownIt from 'markdown-it'
import { describe, test, expect } from 'vitest'
import { tabsPlugin } from '../markdownPlugin'
import dedent from 'ts-dedent'
import container from 'markdown-it-container'

const setup = ({
  withTipContainer = false
}: { withTipContainer?: boolean } = {}) => {
  const md = MarkdownIt()
  md.use(tabsPlugin)
  if (withTipContainer) {
    md.use(container, 'tip')
  }
  return md
}

describe('render correctly', () => {
  const md = setup()

  const inputs = [
    dedent`
      :::tabs
      == Tab "A"
      a
      == Tab "B"
      b
      :::
    `,
    dedent`
      ::::tabs
      === Tab "A"
      a
      === Tab "B"
      b
      ::::
    `,
    dedent`
      :::tabs
      == Tab "A"
      a
      == Tab "B"
      b
    `
  ]

  test.each(inputs)('input', input => {
    const result = md.render(input)
    expect(result).toMatchSnapshot()
  })
})

test('render tab with nested ::: code correctly', () => {
  const md = setup()
  const result = md.render(dedent`
    :::tabs
    == Tab "A"
    a
    \`\`\`js
    const foo = ':::';
    \`\`\`
    == Tab "B"
    b
    :::
  `)
  expect(result).toMatchSnapshot()
})

describe('render tab with nested block correctly', () => {
  const md = setup({ withTipContainer: true })

  const inputs = [
    dedent`
      :::: tabs
      == Tab "A"
      a
      :::tip
      Tip
      :::
      == Tab "B"
      b
      ::::
    `,
    dedent`
      :::: tabs
      == Tab "A"
      a
      :::tip
      Tip
      == Tab "B"
      :::
      == Tab "C"
      c
      ::::
    `,
    dedent`
      :::: tabs
      === Tab "A"
      a
      :::tip
      Tip
      == t
      :::
      === Tab "B"
      b
      ::::
    `
  ]

  test.each(inputs)('input', input => {
    const result = md.render(input)
    expect(result).toMatchSnapshot()
  })
})
