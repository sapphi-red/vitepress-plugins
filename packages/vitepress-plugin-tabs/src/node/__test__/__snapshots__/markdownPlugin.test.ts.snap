// Vitest Snapshot v1

exports[`render correctly > input 1`] = `
"<PluginTabs >
<PluginTabsTab label=\\"Tab &quot;A&quot;\\">
<p>a</p>
</PluginTabsTab>
<PluginTabsTab label=\\"Tab &quot;B&quot;\\">
<p>b</p>
</PluginTabsTab>
</PluginTabs>
"
`;

exports[`render correctly > input 2`] = `
"<PluginTabs >
<PluginTabsTab label=\\"Tab &quot;A&quot;\\">
<p>a</p>
</PluginTabsTab>
<PluginTabsTab label=\\"Tab &quot;B&quot;\\">
<p>b</p>
</PluginTabsTab>
</PluginTabs>
"
`;

exports[`render correctly > input 3`] = `
"<PluginTabs >
<PluginTabsTab label=\\"Tab &quot;A&quot;\\">
<p>a</p>
</PluginTabsTab>
<PluginTabsTab label=\\"Tab &quot;B&quot;\\">
<p>b</p>
</PluginTabsTab>
</PluginTabs>
"
`;

exports[`render tab with nested ::: code correctly 1`] = `
"<PluginTabs >
<PluginTabsTab label=\\"Tab &quot;A&quot;\\">
<p>a</p>
<pre><code class=\\"language-js\\">const foo = ':::';
</code></pre>
</PluginTabsTab>
<PluginTabsTab label=\\"Tab &quot;B&quot;\\">
<p>b</p>
</PluginTabsTab>
</PluginTabs>
"
`;

exports[`render tab with nested block correctly > input 1`] = `
"<PluginTabs >
<PluginTabsTab label=\\"Tab &quot;A&quot;\\">
<p>a</p>
<div class=\\"tip\\">
<p>Tip</p>
</div>
</PluginTabsTab>
<PluginTabsTab label=\\"Tab &quot;B&quot;\\">
<p>b</p>
</PluginTabsTab>
</PluginTabs>
"
`;

exports[`render tab with nested block correctly > input 2`] = `
"<PluginTabs >
<PluginTabsTab label=\\"Tab &quot;A&quot;\\">
<p>a</p>
<div class=\\"tip\\">
<p>Tip</p>
</div>
</PluginTabsTab>
<PluginTabsTab label=\\"Tab &quot;B&quot;\\">
<p>:::</p>
</PluginTabsTab>
<PluginTabsTab label=\\"Tab &quot;C&quot;\\">
<p>c</p>
</PluginTabsTab>
</PluginTabs>
"
`;

exports[`render tab with nested block correctly > input 3`] = `
"<PluginTabs >
<PluginTabsTab label=\\"Tab &quot;A&quot;\\">
<p>a</p>
<div class=\\"tip\\">
<p>Tip
== t</p>
</div>
</PluginTabsTab>
<PluginTabsTab label=\\"Tab &quot;B&quot;\\">
<p>b</p>
</PluginTabsTab>
</PluginTabs>
"
`;
